[CssSelector] Add `:has()` support

franckranaivo opened the pull request:
| Q             | A
| ------------- | ---
| Branch       | 8.1
| Bug fix       | no
| New feature  | yes
| Tickets          | Fix https://github.com/symfony/symfony/issues/40179
| Deprecations | no
| License       | MIT
| Doc PR        |  https://github.com/symfony/symfony-docs/pull/17994

Add support for :has() function.

franckranaivo committed:
[CssSelector] Add `:has()` support

stof commented on src/Symfony/Component/CssSelector/Exception/SyntaxErrorException.php:
```diff
@@ -43,6 +43,11 @@ public static function nestedNot(): self
 
+    public static function nestedHas(): self
+    {
+        return new self('Got nested ::has().');
```
this is `:has()`, not `::has()`

stof commented on src/Symfony/Component/CssSelector/Tests/XPath/TranslatorTest.php:
```diff
@@ -219,6 +219,7 @@ public static function getCssToXPathTestData()
             ['e + f', "e/following-sibling::*[(name() = 'f') and (position() = 1)]"],
             ['e ~ f', 'e/following-sibling::f'],
             ['div#container p', "div[@id = 'container']/descendant-or-self::*/p"],
+            ['div:has(div.foo)', "div[count(descendant-or-self::*[(@class and contains(concat(' ', normalize-space(@class), ' '), ' foo ')) and (name() = 'div')]) > 0]"],
```
we should also have tests covering more complex selectors in `:has()`, as well as relative selectors: `div:has(.foo .bar)`, `div:has(> .foo)`, `div:has(~ foo)`

stof commented on src/Symfony/Component/CssSelector/Node/RelationNode.php:
```diff
@@ -0,0 +1,58 @@
+
+    public function getSpecificity(): Specificity
+    {
+        return $this->selector->getSpecificity()->plus(new Specificity(0, 1, 0));
```
the specificity of `:has()` is not `(0, 1, 0)` but the specificity of the most specific selector in its arguments

stof commented on src/Symfony/Component/CssSelector/Tests/XPath/TranslatorTest.php:
```diff
@@ -219,6 +219,11 @@ public static function getCssToXPathTestData()
+            ['div:has(div.foo)', "div[count(descendant-or-self::div[@class and contains(concat(' ', normalize-space(@class), ' '), ' foo ')]) > 0]"],
+            ['div:has(.foo .bar)', "div[count(descendant-or-self::*[@class and contains(concat(' ', normalize-space(@class), ' '), ' foo ')]/descendant-or-self::*/*[@class and contains(concat(' ', normalize-space(@class), ' '), ' bar ')]) > 0]"],
+            ['div:has(> .foo)', "div[count(descendant-or-self::*/*[@class and contains(concat(' ', normalize-space(@class), ' '), ' foo ')]) > 0]"],
+            ['div:has(~ .foo)', "div[count(descendant-or-self::*/following-sibling::*[@class and contains(concat(' ', normalize-space(@class), ' '), ' foo ')]) > 0]"],
```
`descendant-or-self::*` looks wrong for those combinators. I'm quite sure it should be `self::*` instead

stof replied:
My guess is that you currently parse `div:has(> .foo)` as `div:has(* > .foo)`, which is totally different.

stof commented on src/Symfony/Component/CssSelector/Node/RelationNode.php:
```diff
@@ -0,0 +1,58 @@
+    private NodeInterface $selector;
+    private array $arguments;
+
+    public function __construct(NodeInterface $selector, array $arguments)
```
you should document the type for `$arguments` as `list<NodeInterface>`

stof commented on src/Symfony/Component/CssSelector/Node/RelationNode.php:
```diff
@@ -0,0 +1,58 @@
+ * Represents a "<selector>:has(<arguments>)" node.
+ *
+ * This component is a port of the Python cssselect library,
+ * which is copyright Ian Bicking, @see https://github.com/SimonSapin/cssselect.
```
the right URL is `https://github.com/scrapy/cssselect`

stof commented on src/Symfony/Component/CssSelector/Parser/Parser.php:
```diff
@@ -216,6 +217,21 @@ private function parseSimpleSelector(TokenStream $stream, bool $insideNegation =
+                        throw SyntaxErrorException::nestedHas();
+                    }
+
+                    $arguments[] = $this->parserSelectorNode($stream, true);
```
The Python library has a `parse_relative_selector` function used here. It looks like your implementation is not actually a port of the Python one (or maybe of an outdated version of the Python one)

stof replied:
Btw, the PR introducing it in Python has a bunch of testcases covering it, that you should port (which would avoid adding broken assertions): https://github.com/scrapy/cssselect/pull/115/files

franckranaivo replied:
Indeed, I changed the way the code parses selectors as python library do. Also added relevant test. 

But now, It looks like the parser can't process descendant relation anymore like `div:has(.foo .bar)`. It would throw an error. I'm pretty sure python one neither do. Should we let it like this for now?

stof commented on src/Symfony/Component/CssSelector/Parser/Parser.php:
```diff
@@ -129,15 +129,15 @@ private function parserSelectorNode(TokenStream $stream): Node\SelectorNode
             }
 
-            [$nextSelector, $pseudoElement] = $this->parseSimpleSelector($stream);
+            [$nextSelector, $pseudoElement] = $this->parseSimpleSelector($stream, false);
```
those argument addition look useless

stof commented on src/Symfony/Component/CssSelector/XPath/Extension/ExtensionInterface.php:
```diff
@@ -60,6 +60,13 @@ public function getPseudoClassTranslators(): array;
+     *
+     * @return callable[]
+     */
+    public function getRelativeCombinationTranslators(): array;
```
this must be added using `@method` as adding a new method in an interface is a BC break

stof commented on src/Symfony/Component/CssSelector/XPath/Extension/ExtensionInterface.php:
```diff
@@ -60,6 +60,13 @@ public function getPseudoClassTranslators(): array;
+    /**
+     * Returns relative combinators translators.
+     *
+     * @return callable[]
```
I suggest using `array<string, callable>` to describe the type of keys. Or even better, `array<string, callable(XPathExpr, XPathExpr): XPathExpr>` to describe the callable signature as well.

stof commented on src/Symfony/Component/CssSelector/Tests/XPath/TranslatorTest.php:
```diff
@@ -221,6 +221,10 @@ public static function getCssToXPathTestData()
+            ['e:has(> f)', 'e[./f]'],
+            ['e:has(~ f)', 'e[following-sibling::f]'],
+            ['e:has(+ f)', "e[following-sibling::*[(name() = 'f') and (position() = 1)]]"],
+            ['e:has(f)', 'e[descendant-or-self::f]'],
```
It would be great to add some `:has()` selector to `getHtmlShakespearTestData` or `getHtmlIdsTestData` to also have tests ensuring we get the right elements

franckranaivo replied:
Okay, I added relevant tests on both of these datasets. ~~Anyway, I resolved here (https://github.com/symfony/symfony/pull/50729) the issue linked with `testHtmlIds` tests. Please tell me if I should bring here also changes made there.~~

stof commented on src/Symfony/Component/CssSelector/XPath/Extension/RelationExtension.php:
```diff
@@ -0,0 +1,63 @@
+    public function translateRelationDirectAdjacent(XPathExpr $xpath, XPathExpr $combinedXpath): XPathExpr
+    {
+        return $xpath
+            ->addCondition(sprintf('following-sibling::*[(name() = \'%s\') and (position() = 1)]', $combinedXpath->getElement()));
```
using only `$combinedXpath->getElement()` looks suspicious to me. What if the selector is `:has(+ .baz)` ?

franckranaivo replied:
Ah yes, you're right, but python code only handle `->getElement()` .  as you can see [here](https://github.com/scrapy/cssselect/blob/ddd9784977fca6e7da4439d837e4f510f1f10638/cssselect/xpath.py#L445) .

I will try if I can find a better alternative elsewhere.

franckranaivo commented:
I added relevant tests here and fixed some issues. What else should be done here ?

franckranaivo commented:
I rebased to 6.4 and removed unused method. For Psalm, idk why it is complaining about  unused `getRelativeCombinationTranslators` method as it is used in Translator class (Xpath namespace). Maybe I missed something here, if so please tell me what to do.
  
Should I keep this PR opened?

Friendly ping @stof

Bilge commented:
Why has this stalled?

Bilge commented on src/Symfony/Component/CssSelector/Tests/XPath/TranslatorTest.php:
```diff
@@ -361,6 +367,12 @@ public static function getHtmlIdsTestData()
             [':enabled', ['link-href', 'tag-anchor', 'nofollow-anchor', 'checkbox-unchecked', 'text-checked', 'checkbox-checked', 'area-href']],
             [':disabled', ['checkbox-disabled', 'checkbox-disabled-checked', 'fieldset', 'checkbox-fieldset-disabled']],
             [':checked', ['checkbox-checked', 'checkbox-disabled-checked']],
+            ['div:has(> span)', ['foobar-div']],
+            ['ol:has(> .c)', ['first-ol']],
+            ['div:has( .a)', ['outer-div']],
+            ['a:has(~ a)', ['name-anchor', 'tag-anchor']],
+            ['ol:has(+ p)', ['first-ol']],
+            ['a:has(+ .b)', ['nofollow-anchor']],
```
* Looks like this is missing a test case for multiple selectors. When `has` has multiple selectors, e.g. `:has(foo, bar)`, it will match if either `foo` **or** `bar` are present.
* We should also test for `:has(:not(foo, bar))` which will only match if `foo` **and** `bar` are not present.
---
Source: https://drafts.csswg.org/selectors-4/#relational

franckranaivo replied:
Good catch, I will add these right away.

franckranaivo replied:
An update for the missing test case @Bilge mentioned above. 

I just saw that @stof has already posted an issue on Cssselect (python parser). Their version does not support relative selector list [here](https://github.com/scrapy/cssselect/issues/138) and remains unresolved until now. I guess this is why this PR stalled.

I will try to find another way.

Bilge commented:
Where are we with this? Is this ready to merge?

nicolas-grekas reviewed (approved):
Known limitations (acceptable for v1, matches Python upstream):

No attribute selectors inside `:has()` (e.g. `:has([data-x])`)
No compound descendant selectors (e.g. `:has(.foo .bar)`)
No selector lists (e.g. `:has(.foo, .bar)`)
No nested pseudo-classes (e.g. `:has(:not(.foo))`)

nicolas-grekas commented:
Can you have a look please?

nicolas-grekas commented:
Was easy enough to add, see b20df88c246

Bilge commented:
A `NotSupportedException` would be a slight improvement but otherwise this seems good enough.
